### DISCUSSION TITLE
Inverse Rendering for EM fields

### DISCUSSION THREAD

**Discussion by pingpongballz** [mitsuba-renderer/mitsuba3#1711] [Q&A]
Hi all, I know this isn't my first time here 😅, but was wondering if this certain idea is feasible in the context of inverse rendering

Please pardon my lack of knowledge, I come from an electrical engineering background 😅 

Basically, I have an antenna at a certain location $r$ that picks up scattered electromagnetic fields from an object. 

<img width="310" height="193" alt="image" src="https://github.com/user-attachments/assets/63651b51-790f-4ec2-9c78-c6ba2deb03a6" />

The transmitter shoots rays, and it radiates off the surface. The intensity of the radiation as seen by the receiver is as follows.
$\vec{E_s} = 2ik\iint_{rays} \hat{r} \times (\hat{r} \times ( \hat{n} \times \hat{H_i} ) e^{ik(t - \hat{r} \cdot \vec{r'})}ds'$, where $\hat{r}$ is the unit camera direction, $\vec{r'}$ is the hit location of the rays, and $\hat{H_i}$ is the initial ray polarisation. The result is a single 3D complex vector.

From what I've read in literature, using the Reynold's transport theorem, I can possibly decompose the derivative of this integral as $\frac{\partial}{\partial \pi} \vec{E_s} = \iint \frac{\partial}{\partial \pi} \vec{F}(r', \pi) ds' + \oint \Delta{\vec{F}} (\frac{\partial \vec{r'}}{\partial \pi} \cdot \hat{n}) dl$, where $\vec{F} = 2ik\hat{r} \times (\hat{r} \times ( \hat{n} \times \hat{H_i} ) e^{ik(t - \hat{r} \cdot \vec{r'})}$.

Then I can possibly use mitsuba's direct projective integrators to perform inverse rendering in the context of this? 

I tried looking into Sionna RT, but it is a completely different application use case.

If it can be done, do I have to modify the sensors to behave more like the receive antenna, or maybe modify the BSDFs? Because the scattered field emanating from the surfaces do not really behave like the reflection types given in mitsuba.

Once again, thank you for your time 😄

**Comment by njroussel** [comment]
Hi @pingpongballz

> Then I can possibly use mitsuba's direct projective integrators to perform inverse rendering in the context of this?

Yes, that's the sort of integrals that we solve in inverse rendering. 

> If it can be done, do I have to modify the sensors to behave more like the receive antenna, or maybe modify the BSDFs? Because the scattered field emanating from the surfaces do not really behave like the reflection types given in mitsuba.

I'm not so sure about this, because I don't quite know what the differences are. I imagine that these would indeed both need to be changed to correctly reflect how the scattered field behaves. One important note is that we don't have any tools in Mitsuba to efficiently handle differentiation of polarized light.

**Reply by pingpongballz** [reply]
Hi @njroussel,

Thank you! Are there any methods or functions built into mitsuba to compute the term $\oint \Delta{\vec{F}} (\frac{\partial \vec{r'}}{\partial \pi} \cdot \hat{n}) dl$?

Once again, thank you 😀

**Reply by njroussel** [reply]
Currently, in Mitsuba, only the `*_projective` integrators are capable of handling discontinuities in the integrand. Most of the relevant code is in this method: https://github.com/mitsuba-renderer/mitsuba3/blob/master/src/python/python/ad/projective.py#L532. This whole logic is fairly complex, you might actually want to start at the very top of the callstack and walk your way down, i.e here: https://github.com/mitsuba-renderer/mitsuba3/blob/master/src/python/python/ad/integrators/common.py#L1013-L1053

**Reply by pingpongballz** [reply]
Hi @njroussel,

Thank you. 

I was thinking for a näive first implementation, I can obtain the directly visible boundary by obtaining a mask of rays that hit or miss a scene, and then definining the boundary as the contour of the mask. From there, I just project the rays from the contour to the scene to obtain my boundary.

I then compute the boundary integral by summing up the fields on the boundary to get my boundary derivative, then add it to the interior derivative obtained by AD to get my final derivative.

Is this flow a logical flow for this type of problem? Pardon me, I'm new to this 😅

**Reply by njroussel** [reply]
> I can obtain the directly visible boundary by obtaining a mask of rays that hit or miss a scene, and then definining the boundary as the contour of the mask.

I imagine this mask is an image? Depending on the scene this might work for a high resolution mask. However, more generally, you want your ray to be exactly on the shape's boundary/silhouette. If you're just using pixel boundaries you'll naturally have small bias as the pixel has a certain width. Ultimately, this would just mean that some of the terms in your boundary integral are slightly wrong. 

It might still be a good approximation and be sufficient to optimize the inverse problems you're considering.

**Reply by pingpongballz** [reply]
Hi @njroussel 

I see. So the flow I proposed would work in theory? Other than the boundary integral being slightly wrong from the pixel approximation?

**Comment by pingpongballz** [comment]
Hi @njroussel,

I just have one more question, if you are ok with that. 
$\frac{\partial}{\partial \pi} \vec{E_s} = \iint \frac{\partial}{\partial \pi} \vec{F}(r', \pi) ds' + \oint \Delta{\vec{F}} (\frac{\partial \vec{r'}}{\partial \pi} \cdot \hat{n}) dl$
I understand that the first term can be computed by autodiff, and I have managed to compute the second boundary term by sampling the silhouette. 
The question is, how do I link the second term such that the dr.backward() and opt.step() methods can incorporate the second boundary term into the gradient optimisation?

**Reply by njroussel** [reply]
This is where you'll need to tap into some of the more precise AD functions of the Dr.Jit API.

One option is to follow the architecture we have for the `PSIntegrator` that I shared in one of my earlier comments. In short, it relies on a Dr.Jit custom operation ([documentation](https://drjit.readthedocs.io/en/latest/reference.html#drjit.CustomOp)) that defines how gradients are progagated backward from a `mi.render()` call ([source code definition](https://github.com/mitsuba-renderer/mitsuba3/blob/master/src/python/python/util.py#L394-L525)). In the `backward` definition for the `RenderOp`, there's a call to `render_backward` which simply [sums both components together](https://github.com/mitsuba-renderer/mitsuba3/blob/master/src/python/python/ad/integrators/common.py#L911-L1011).

Another option is to call `backward()` twice (once for each component) and let the gradients accumulate on your parameters. For the first term, it's as you would do normally. For the second term, it's a bit trickier. You'll want to do something like this: 

```python
first_term = interior_term()
reference_value = get_ref()
loss = loss_func(first_term, reference_value)

dr.backward(loss) # parameters have received some gradients now
# Depending on how your code is setup, you potentially might need to specify `flags=dr.ADFlags.ClearVertices`)


second_term = boundary_term()
# This operations replaces the AD graph of the `first_term` with the one of your second term
# This is required such that the "primal" value (forward simulation value) still matches, but the gradients will be propagated as defined by the boundary term
second_term = dr.replace_grad(first_term, second_term)

dr.backward(loss) # New gradients will be accumulated onto the existing one at the parameters being differentiated.
```

**Reply by pingpongballz** [reply]
Hi @njroussel,

If I'm understanding this correctly, $\frac{\partial{L}}{\partial{\pi}} = \frac{\partial{L}}{\partial{E_s}} \frac{\partial{E_s}}{\partial{\pi}} = \frac{\partial{L}}{\partial{E_s}} \iint \frac{\partial}{\partial \pi} \vec{F}(r', \pi) ds' +  \frac{\partial{L}}{\partial{E_s}} \oint \Delta{\vec{F}} (\frac{\partial \vec{r'}}{\partial \pi} \cdot \hat{n}) dl$ . 

The first term of this loss derivative, $\frac{\partial{L}}{\partial{E_s}} \iint \frac{\partial}{\partial \pi} \vec{F}(r', \pi) ds'$, can be obtained by ```dr.backward(loss)```. 

Then afterward we replace the gradient value of ```first_term``` with our boundary term $\oint \Delta{\vec{F}} (\frac{\partial \vec{r'}}{\partial \pi} \cdot \hat{n}) dl$ . In my implementation, I used a method to find the boundary term almost identical to [this](https://github.com/mitsuba-renderer/mitsuba3/blob/82f8f2650e23ffa226aa7225c43d80586d76b340/src/python/python/ad/integrators/common.py#L1131).

Lastly, we do ```dr.backward(loss)``` to propagate the new gradients.

Is this understanding correct? What I find confusing is why are we defining the loss here as a function of the first term $\iint \frac{\partial}{\partial \pi} \vec{F}(r', \pi) ds'$, when the first term is a differential? Likewise, same for the second term $\oint \Delta{\vec{F}} (\frac{\partial \vec{r'}}{\partial \pi} \cdot \hat{n}) dl$.

Once again, apologies if this is a trivial concept.

**Reply by njroussel** [reply]
> What I find confusing is why are we defining the loss here as a function of the first term, when the first term is a differential?

Not exactly, I kind of abused the naming of the variables in the snippet I shared. Unless you were explicitly computing the derivative, this term in the code just has the value of the usual (forward) simulation. Only once `dr.backward` is called, are you actually computing the derivatives. Effectively the `first_term` in the code is just the result of `mi.render(scene, ...)`. We still need to compute the loss before callling `dr.backward()`, which will then produce the derivative.

**Reply by pingpongballz** [reply]
Hi @njroussel,

I see. But since the second term is effectively part of the derivative and not exactly part of the forward simulation result, how is it managed?

**Reply by njroussel** [reply]
You're correct. Similarly to the first term, the second term's variable is not the derivative. The second term's variable was specifically assigned with code that will produce the correct derivative when "autodiffed". If you look at the code you linked ([this snippet](https://github.com/mitsuba-renderer/mitsuba3/blob/82f8f2650e23ffa226aa7225c43d80586d76b340/src/python/python/ad/integrators/common.py#L1131)) the comment is misleading as it's not the derivative **yet**, but some primal value that will produce the correct derivative when differentiated. 

As you said, the second term is not really a part of the forward simulation. That's why we do the `second_term = dr.replace_grad(first_term, second_term)` trick: we're effectively tying it into the simulation without changing the primal value, but just the way in which gradients will be propagated back.

**Comment by pingpongballz** [comment]
Hi,

Sorry for reviving this thread, but below is my implementation:

```Python
reference_value = get_ref()
first_term = interior_term()
second_term = boundary_term()

total_term = first_term + dr.replace_grad(0.0 * first_term, second_term)

loss = loss_func(total_term, reference_value)

dr.backward(loss)

```

Is this implementation correct with use of drjit functions?
